File: src/syncloud_platform/tools/disk/lsblk.py
from subprocess import check_output
import re
from syncloud_app import logger

PARTTYPE_EXTENDED = '0x5'


class Lsblk:

    def __init__(self, platform_config, path_checker):
        self.platform_config = platform_config
        self.path_checker = path_checker
        self.log = logger.get_logger('lsblk')

    def available_disks(self, lsblk_output=None):
        if not lsblk_output:
            lsblk_output = check_output('lsblk -Pp -o NAME,SIZE,TYPE,MOUNTPOINT,PARTTYPE,MODEL', shell=True)
        disks = []
        disk = None
        for line in lsblk_output.splitlines():
            match = re.match(
                r'NAME="(.*)" SIZE="(.*)" TYPE="(.*)" MOUNTPOINT="(.*)" PARTTYPE="(.*)" MODEL="(.*)"',
                line.strip())

            lsblk_entry = LsblkEntry(match.group(1), match.group(2), match.group(3),
                                     match.group(4), match.group(5), match.group(6).strip())

            if lsblk_entry.type in ('disk', 'loop'):
                disk_name = lsblk_entry.model.split(' ')[0]
                disk = Disk(disk_name)
                if lsblk_entry.type == 'loop':
                    self.add_partition(disk, lsblk_entry)
                disks.append(disk)

            elif lsblk_entry.type == 'part':
                self.add_partition(disk, lsblk_entry)

        disks_with_partitions = [d for d in disks if d.partitions]
        return disks_with_partitions

    def add_partition(self, disk, lsblk_entry):
        mountable = False
        mount_point = lsblk_entry.mount_point
        if not lsblk_entry.is_extended_partition():
            if not mount_point or mount_point == self.platform_config.get_external_disk_dir():
                mountable = True

        if lsblk_entry.is_boot_disk():
            mountable = False
        active = False
        if mount_point == self.platform_config.get_external_disk_dir() \
                and self.path_checker.external_disk_link_exists():
            active = True
        if mountable:
            disk.partitions.append(Partition(lsblk_entry.size, lsblk_entry.name, mount_point, active))

    def is_external_disk_attached(self, lsblk_output=None, disk_dir=None):
        if not disk_dir:
            disk_dir = self.platform_config.get_external_disk_dir()
        for disk in self.available_disks(lsblk_output):
            for partition in disk.partitions:
                if partition.mount_point == disk_dir:
                    self.log.info('external disk is attached')
                    return True
        self.log.info('external disk is detached')
        return False


class LsblkEntry:
    def __init__(self, name, size, fs_type, mount_point, part_type, model):
        self.name = name
        self.size = size
        self.type = fs_type
        self.mount_point = mount_point
        self.parttype = part_type
        self.model = model

    def is_extended_partition(self):
        return self.parttype == PARTTYPE_EXTENDED

    def is_boot_disk(self):
        return '/dev/mmcblk0' in self.name


class Partition:
    def __init__(self, size, device, mount_point, active):
        self.size = size
        self.device = device
        self.mount_point = mount_point
        self.active = active

    def __str__(self):
        return '{0}, {1}, {2}, {3}'.format(self.device, self.size, self.mount_point, self.active)


class Disk:
    def __init__(self, name):
        self.partitions = []
        self.name = name

    def __str__(self):
        return '{0}: {1}'.format(self.name, ','.join(map(str, self.partitions)))

File: backend/storage/lsblk.go
package storage

import (
	"fmt"
	"github.com/syncloud/platform/cli"
	"github.com/syncloud/platform/storage/model"
	"go.uber.org/zap"
	"regexp"
	"sort"
	"strings"
)

type Lsblk struct {
	systemConfig Config
	pathChecker  Checker
	executor     cli.Executor
	logger       *zap.Logger
}

type Config interface {
	ExternalDiskDir() string
}

type ByDevice []model.Disk

func (a ByDevice) Len() int           { return len(a) }
func (a ByDevice) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByDevice) Less(i, j int) bool { return a[i].Device < a[j].Device }

func NewLsblk(config Config, pathChecker Checker, executor cli.Executor, logger *zap.Logger) *Lsblk {
	return &Lsblk{
		systemConfig: config,
		pathChecker:  pathChecker,
		executor:     executor,
		logger:       logger,
	}
}

func (l *Lsblk) AvailableDisks() ([]model.Disk, error) {
	var disks []model.Disk

	allDisks, err := l.AllDisks()
	if err != nil {
		return nil, err
	}
	for _, disk := range allDisks {
		if disk.IsAvailable() {
			disks = append(disks, disk)
		}
	}
	return disks, nil
}

func (l *Lsblk) parseLsblkOutput() ([]model.LsblkEntry, error) {
	var entries []model.LsblkEntry

	lsblkOutputBytes, err := l.executor.CombinedOutput("lsblk", "-Pp", "-o", "NAME,SIZE,TYPE,MOUNTPOINT,PARTTYPE,FSTYPE,MODEL,UUID")
	if err != nil {
		return nil, err
	}
	lsblkOutput := string(lsblkOutputBytes)
	lsblkLines := strings.Split(lsblkOutput, "\n")
	for _, rawLine := range lsblkLines {
		line := strings.TrimSpace(rawLine)
		if line == "" {
			continue
		}
		r, err := regexp.Compile(`NAME="(.*)" SIZE="(.*)" TYPE="(.*)" MOUNTPOINT="(.*)" PARTTYPE="(.*)" FSTYPE="(.*)" MODEL="(.*)" UUID="(.*)"`)
		if err != nil {
			return nil, err
		}

		match := r.FindStringSubmatch(line)
		mountPoint := match[4]

		entries = append(entries, model.LsblkEntry{
			Name:       match[1],
			Size:       match[2],
			DeviceType: match[3],
			MountPoint: mountPoint,
			PartType:   match[5],
			FsType:     match[6],
			Model:      strings.TrimSpace(match[7]),
			Active:     l.isActive(mountPoint),
			Uuid:       match[8],
		})

	}

	return entries, nil
}

func (l *Lsblk) extractActiveUuid(entries []model.LsblkEntry) map[string]bool {
	uuids := make(map[string]bool)
	for _, entry := range entries {
		if entry.Active && entry.Uuid != "" {
			uuids[entry.Uuid] = true
		}
	}
	return uuids
}

func (l *Lsblk) AllDisks() ([]model.Disk, error) {
	disks := make(map[string]*model.Disk)
	entries, err := l.parseLsblkOutput()
	activeUuids := l.extractActiveUuid(entries)
	if err != nil {
		return nil, err
	}
	for _, entry := range entries {
		if entry.IsSupportedType() && entry.IsSupportedFsType() {
			device := entry.Name
			diskName := entry.Model
			active := entry.Active
			if !active {
				active = activeUuids[entry.Uuid]
			}
			disk := model.NewDisk(diskName, device, entry.Size, active, entry.Uuid, entry.MountPoint, []model.Partition{})
			if entry.IsRaid() {
				disk.Name = entry.DeviceType
				partition := l.createPartition(entry)
				disk.AddPartition(partition)
			}

			disks[device] = disk

		} else if entry.DeviceType == "part" {
			partition := l.createPartition(entry)
			parentDevice, err := entry.ParentDevice()
			if err != nil {
				return nil, err
			}

			if _, ok := disks[parentDevice]; ok {
				disk := disks[parentDevice]
				disk.AddPartition(partition)
			}
		}
	}

	var results []model.Disk
	for _, disk := range disks {
		results = append(results, *disk)
	}
	sort.Sort(ByDevice(results))
	return results, nil
}

func (l *Lsblk) createPartition(lsblkEntry model.LsblkEntry) model.Partition {
	return model.Partition{
		Size:       lsblkEntry.Size,
		Device:     lsblkEntry.Name,
		MountPoint: lsblkEntry.MountPoint,
		Active:     lsblkEntry.Active,
		FsType:     lsblkEntry.GetFsType()}
}

func (l *Lsblk) isActive(mountPoint string) bool {
	active := false
	if mountPoint == l.systemConfig.ExternalDiskDir() && l.pathChecker.ExternalDiskLinkExists() {
		active = true
	}
	return active
}

func (l *Lsblk) FindPartitionByDevice(device string) (*model.Partition, error) {
	disks, err := l.AllDisks()
	if err != nil {
		return nil, err
	}
	for _, disk := range disks {
		for _, partition := range disk.Partitions {
			if partition.Device == device {
				l.logger.Info("partition found")
				return &partition, nil
			}
		}
	}
	return nil, fmt.Errorf("unable to find device: %s", device)
}
